runtime/src/accounts_db.rs

```diff
@@ -2723,6 +2804,131 @@ pub mod tests {
+        );
+
+        // Executable may not be modified
+        let mut account_modified = account.clone();
+        account_modified.executable = true;
+        assert_ne!(
+            hash,
+            AccountsDB::hash_frozen_account_data(&account_modified)
+        );
+    }
+
+    #[test]
+    fn test_frozen_account_lamport_increase() {
+        let frozen_pubkey =
+            Pubkey::from_str("My11111111111111111111111111111111111111111").unwrap();
+        let mut db = AccountsDB::new(Vec::new());
+
+        let mut account = Account::new(1, 42, &frozen_pubkey);
+        db.store(0, &[(&frozen_pubkey, &account)]);
+
+        let ancestors = vec![(0, 0)].into_iter().collect();
+        db.freeze_accounts(&ancestors, &[frozen_pubkey]);
+
+        // Store with no account changes is ok
+        db.store(0, &[(&frozen_pubkey, &account)]);
+
+        // Store with an increase in lamports is ok
+        account.lamports += 1;
+        db.store(0, &[(&frozen_pubkey, &account)]);
+
+        // Store with an decrease that does not go below the frozen amount of lamports is tolerated
+        account.lamports -= 1;
+        db.store(0, &[(&frozen_pubkey, &account)]);
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "Frozen account My11111111111111111111111111111111111111111 modified.  Lamports decreased from 1 to 0"
+    )]
+    fn test_frozen_account_lamport_decrease() {
+        let frozen_pubkey =
+            Pubkey::from_str("My11111111111111111111111111111111111111111").unwrap();
+        let mut db = AccountsDB::new(Vec::new());
+
+        let mut account = Account::new(1, 42, &frozen_pubkey);
+        db.store(0, &[(&frozen_pubkey, &account)]);
+
+        let ancestors = vec![(0, 0)].into_iter().collect();
+        db.freeze_accounts(&ancestors, &[frozen_pubkey]);
+
+        // Store with a decrease below the frozen amount of lamports is not ok
+        account.lamports -= 1;
+        db.store(0, &[(&frozen_pubkey, &account)]);
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "Unable to freeze an account that does not exist: My11111111111111111111111111111111111111111"
+    )]
+    fn test_frozen_account_nonexistent() {
+        let frozen_pubkey =
+            Pubkey::from_str("My11111111111111111111111111111111111111111").unwrap();
+        let mut db = AccountsDB::new(Vec::new());
+
+        let ancestors = vec![(0, 0)].into_iter().collect();
+        db.freeze_accounts(&ancestors, &[frozen_pubkey]);
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "Frozen account My11111111111111111111111111111111111111111 modified.  Hash changed from 8wHcxDkjiwdrkPAsDnmNrF1UDGJFAtZzPQBSVweY3yRA to JdscGYB1uczVssmYuJusDD1Bfe6wpNeeho8XjcH8inN"
+    )]
+    fn test_frozen_account_data_modified() {
+        let frozen_pubkey =
+            Pubkey::from_str("My11111111111111111111111111111111111111111").unwrap();
+        let mut db = AccountsDB::new(Vec::new());
```

Comment:
Can we add the test case I described. All these tests only operate on updates on a single slot which is a bit worrying.

Reply:
Oh I missed that comment I guess.  What test case should I add?

Reply:
If I have an account which has 5 lamports in slot 0, then I increase it to 10 lamports in slot 1, then I have slot 2 with slot 0 as parent, and I increase it to 7 lamports there.

Reply:
Oh so I reworked the code so that this doesn't matter anymore.   The lamports are frozen at AccountsDb initialization, and only a reduction below the frozen amount will be flagged.  In your example, any amount over 5 lamports would be fine and the amount can even reduce *as long as it's always over 5*.

Reply:
Sure, but it would be nice to have one case that tests some cross-slot logic and that a future refactoring doesn't break that behavior.

Reply:
Ugh some local-cluster thing?  This sounds painful

Reply:
No I was thinking just unit test like these

Reply:
Ah, yep.  I added 489f1de33ca5f7f8012e499b6d8daa2cd51a6fad and rebased on your RwLock removal

Reply:
Sweet, thanks.